pkg/operators/wasm/datasource.go

```diff
@@ -0,0 +1,525 @@
+// Copyright 2024 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wasm
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/tetratelabs/wazero"
+	wapi "github.com/tetratelabs/wazero/api"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/datasource"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+)
+
+func (i *wasmOperatorInstance) addDataSourceFuncs(env wazero.HostModuleBuilder) {
+	exportFunction(env, "newDataSource", i.newDataSource,
+		[]wapi.ValueType{
+			wapi.ValueTypeI64, // DataSourceName
+			wapi.ValueTypeI32, // DataSourceType
+		},
+		[]wapi.ValueType{wapi.ValueTypeI32}, // DataSource
+	)
```

Comment:
What would be the ideal solution, but I haven't been able to return multiple values from one of this functions. I have a reproducer in https://github.com/mauriciovasquezbernal/wasm-multi-return, I'm not sure where the limitation is. We have the same issue on `fieldAccessorGet`, we can't differentiate between returning 0 because there was an error, or returning 0 because that's the value of the field. @alban proposed an alternative in https://github.com/inspektor-gadget/inspektor-gadget/pull/2818#discussion_r1616141367, but I'm not sure it's worth it.

Reply:
The https://github.com/tetratelabs/wazero/tree/main/examples/multiple-results example is very interesting, unfortunately the guest code is implemented directly in wasm, and not in Golang as we want. It seems to me the "limitation" is in tinygo, but I haven't been able to understand the whole picture.
- https://github.com/tinygo-org/tinygo/issues/2512
- https://github.com/tinygo-org/tinygo/issues/3254

 AFAIU it's not a proper limitation, but a design decision to keep compatibility (with programming languages that don't support multiple return values?). The [documentation](https://tinygo.org/docs/concepts/compiler-internals/calling-convention/) states:

> The WebAssembly target does not return variables directly that cannot be handled by JavaScript (see above about i64, also struct, i64, multiple return values, etc). Instead, they are stored into a pointer passed as the first parameter by the caller.
> 
> This is the calling convention as implemented by LLVM, with the extension that i64 return values are returned in the same way as aggregate types.

So probably that explains the behavior I'm seeing. 

I'd like to explore this more, together with Alban's idea on https://github.com/inspektor-gadget/inspektor-gadget/pull/2818#discussion_r1616141367, but I think we can move forward with this PR and do that in a following one. 

Any thoughts about this @alban @flyth?

Reply:
Opened #3071 to track this